OpenStack Destroy Cluster Support

hardys opened the pull request:
Adds initial destroy support for OpenStack, currently this only considers Server resources, network resources will be considered either via an update to this PR or follow-up PR's, whichever is preferable.

hardys commented:
Ok I added the vendored dependencies, and rebased on the latest logger changes - this is now working well for me to delete the servers.

Do we want to land this and incrementally add the remaining resources (network/subnet/port/router), or should I keep pushing to this one PR?

hardys commented:
Also I removed WIP as this now works, albeit only for the server resources.

wking commented on pkg/destroy/openstack/openstack_deprovision.go:
```diff
@@ -0,0 +1,187 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+// OpenStackFilter holds the key/value pairs for the tags we will be matching against
```
nit: This is missing trailing period.

Some of your other godocs comments are as well.  Obviously not the end of the world, but should be easy to fix ;).

And we probably have a number of godocs in master with the same issue; PRs welcome :).

hardys replied:
Ah thanks, I'll go through and check - is there a way to build the godocs locally so I can check the format before pushing?

hardys commented:
Note I spent some time today looking at adding tags support for the network resources - unfortunately neither gophercloud or terraform-provider-openstack yet support tags for neutron resources, so looks like I'll have to add those before I can make progress, unless anyone has other suggestions on how we might identify the resources for removal?

russellb commented:
Can we abuse the resource names as a way to tag them for removal?

hardys commented:
> Can we abuse the resource names as a way to tag them for removal?

Yeah I guess so, and the libvirt destroy code does that (looks for a prefix), the names would end up pretty ugly though (particularly since the installer isn't idempotent so it could easily reuse a cluster name on each run creating duplicate resources, sigh).

I did look at gophercloud today and writing the tags looks fairly simple ref https://github.com/gophercloud/gophercloud/pull/1259 (WIP)

flaper87 commented:
(it won't succeed but FYI)

/test e2e-openstack

hardys commented:
Ok I'll mark this WIP although the top server commit could potentially land now - the network patches will all depend on terraform-provider-openstack changes that are under review ref https://github.com/terraform-providers/terraform-provider-openstack/issues/453

If anyone wants to help test you can pull the staged tree with all commits at https://github.com/hardys/terraform-provider-openstack/commits/neutron_tags_wip

hardys commented:
Note this was blocked on gophercloud patches (all now landed and included in the vendored files in the commits I just pushed), and terraform-provider-openstack patches (under review as mentioned).

I've rebased this and the server/port patches work for me, so I'll go ahead and push support for the other resources so hopefully this will be in good shape when the terraform-provider-openstack patches merge (hopefully next week but we'll see how it goes).

hardys commented:
Ok removed WIP as this is now ready for review - note it does still require latest trunk terraform-provider-openstack as my PRs all merged but have not yet been released.

flaper87 commented:
/test images
/test e2e-aws

hardys commented:
@flaper87 I rebased this on https://github.com/openshift/installer/pull/588 as requested

flaper87 commented:
/test e2e-openstack

flaper87 commented:
@hardys thanks! I've triggered the `e2e-openstack` job. You should be able to see the `teardown` container in the `e2e-openstack` pod calling the destroy code (instead of saying "openstack doesn't have a destroyer")

flaper87 commented:
> @flaper87 thanks - looks like we fail to run terraform because this requires latest trunk, e.g we need a new release I guess?
> 
> ```
> Error: module.topology.openstack_networking_network_v2.openshift-private: : invalid or unknown key: tags
> ```
> Edit to clarify, we need the patches that landed via [terraform-providers/terraform-provider-openstack#453](https://github.com/terraform-providers/terraform-provider-openstack/issues/453) which are on terraform-provider-openstack master but not yet tagged in any release

Correct! I wonder if there's a way for us to pin/vendor the `terraform-provider-openstack` rather than depending on upstream releases. At least during this phase of development. @wking any idea if this is possible?

wking commented:
> Correct! I wonder if there's a way for us to pin/vendor the `terraform-provider-openstack`...

Do they tag releases?  If so, maybe like ac5aeed6db51a3674220a72fe893db3072b5ea62?

flaper87 commented:
> > Correct! I wonder if there's a way for us to pin/vendor the `terraform-provider-openstack`...
> 
> Do they tag releases? If so, maybe like [ac5aeed](https://github.com/openshift/installer/commit/ac5aeed6db51a3674220a72fe893db3072b5ea62)?

They tag releases and this is what we're doing now. So, I think we just need to wait for them to tag another one.

I was wondering if we could just get a commit from master and use that but it sounds like it's not possible. Thanks :smile:

hardys commented:
@flaper87 @tomassedovic as discussed I rebased this on the latest #588 and squashed to two commits, the vendor additions for gophercloud and the destroy support itself.

You can test it by deploying as normal, then doing something like:

  ./bin/openshift-install --log-level=debug --dir rhcos destroy cluster

hardys commented:
@flaper87 @tomassedovic as discussed I rebased this on the latest #588 and squashed to two commits, the vendor additions for gophercloud and the destroy support itself.

You can test it by deploying as normal, then doing something like:

  ./bin/openshift-install --log-level=debug --dir rhcos destroy cluster

tomassedovic commented on cmd/openshift-install/destroy.go:
```diff
@@ -9,6 +9,7 @@ import (
 	"github.com/openshift/installer/pkg/destroy"
 	"github.com/openshift/installer/pkg/destroy/bootstrap"
 	_ "github.com/openshift/installer/pkg/destroy/libvirt"
+	_ "github.com/openshift/installer/pkg/destroy/openstack"
```
I don't see this package anywhere (and my build fails because of that). Is it possible you've left it out in the squash/rebase?

hardys replied:
@tomassedovic oops, yes you're right, the new package got lost due to the most recent rebase, I'll re-add it and rebase again now #588 landed, thanks for spotting it!

hardys commented:
Marking WIP again since I need to address the rebase issue and we're still blocked on a new release of terraform-provider-openstack, hopefully that will happen this week

hardys commented:
https://github.com/terraform-providers/terraform-provider-openstack/releases/tag/v1.12.0 has been released so I'll adjust the commit message and remove the WIP

flaper87 commented:
/hold cancel
/lgtm

flaper87 commented:
/ok-to-test
